python/ray/air/_internal/torch_utils.py

```diff
@@ -481,26 +504,37 @@ def move_tensors_to_device(
         device: The device to move tensors to. If None, tensors are not moved.
         non_blocking: If True, perform device transfer without forcing a
             synchronization.
+        pin_memory: Whether to pin the memory of the tensors. Defaults to False.
 
     Returns:
         The input tensors moved to the specified device
     """
     if device is None:
         return batch
 
+    def to_device(tensor: torch.Tensor) -> torch.Tensor:
+        tensor = tensor_pin_memory_if_needed(tensor, pin_memory)
+
+        return tensor.to(
+            device, non_blocking=is_tensor_non_blocking_transfer(tensor, non_blocking)
+        )
```

Comment:
I think pinning memory should be separated into a different stage of [the iterator pipeline](https://github.com/ray-project/ray/blob/master/python/ray/data/_internal/block_batching/iter_batches.py#L123) -- after collate and before finalize (to).

Otherwise:

> Using `tensor.pin_memory().to(device, non_blocking=True)` can be up to twice as slow as a straightforward tensor.to(device).

Refer to https://docs.pytorch.org/tutorials/intermediate/pinmem_nonblock.html#pin-memory and https://docs.pytorch.org/tutorials/intermediate/pinmem_nonblock.html#additional-considerations

`tensor.pin_memory().to(device, non_blocking=True)` is what `move_tensors_to_device(..., pin_memory=True)` does during the finalize stage.